File: Array/linchuanxin/Algorithm/day31_0701面试题 02.02. 返回倒数第 k 个节点/day31_leetcode_interview0202.java
package day31;

public class day31_leetcode_interview0202 {

    public class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
    }

    //k=2
    //null-->1-->2-->3-->4-->5
    //           p



                       //2
    //null<--1<--2<--3<--4<--5
                            //2
    //null<--1<--2<--3<--4<--5<--6
    //                  pre  cur

    public int kthToLast(ListNode head, int k) {
        ListNode fast = head;
        ListNode slow = head;

        while (k!=0){
            k--;
            fast = fast.next;
        }

        while (fast!=null){
            fast = fast.next;
            slow = slow.next;
        }
        return slow.val;

    }
}

File: Array/linchuanxin/Algorithm/day31_0701面试题 02.02. 返回倒数第 k 个节点/day31_leetcode_interview0202.kt
package day31


  class ListNode(var `val`: Int) {
      var next: ListNode? = null
  }
class Solution {
    fun kthToLast(head: ListNode?, k: Int): Int {
        var fast = head
        var slow = head
        var kk = k
        while (kk!=0){
            kk -= 1;
            fast = fast?.next
        }

        while (fast!=null){
            fast = fast.next
            slow = slow?.next
        }

        return slow!!.`val`

    }
}